enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,264 @@
+approvers:
+  - "@yuqi-zhang"
+api-approvers: 
+  - None
+creation-date: 2023-10-16
+last-updated: 2022-11-17
+tracking-link:
+  - https://issues.redhat.com/browse/MCO-589
+see-also:
+replaces: 
+  - https://github.com/openshift/enhancements/pull/368
+superseded-by: 
+  - https://github.com/openshift/enhancements/pull/201
+---
+
+# Managing boot images via the MCO
+
+## Summary
+
+This is a proposal to manage bootimages via the `Machine Config Operator`(MCO), leveraging some of the [pre-work](https://github.com/openshift/installer/pull/4760) done as a result of the discussion in [#201](https://github.com/openshift/enhancements/pull/201). This feature will only target standalone OCP installs. It will also be user opt-in and is planned to be released behind a feature gate.
+
+For Installer Provisioned Infrastructure(IPI) clusters, the end goal is to create a mechanism that can:
+- update the boot images references in `MachineSets` to the latest in the payload image
+- ensure stub ignition referenced in each `Machinesets` is in spec 3 format
+
+For User Provisioned Infrastructure(UPI) clusters, this end goal is to create a document(KB or otherwise) that a cluster admin would follow to update their boot images.
+
+
+## Motivation
+
+Currently, bootimage references are [stored](https://github.com/openshift/installer/blob/1ca0848f0f8b2ca9758493afa26bf43ebcd70410/pkg/asset/machines/gcp/machines.go#L204C1-L204C1) in a `MachineSet` by the openshift installer during cluster bringup and is thereafter not managed. These boot image references are not updated on an upgrade, so any node scaled up using it will boot up with the original “install” bootimage. This has caused a myriad of issues during scale-up due to this version skew, when the nodes attempt the final pivot to the release payload image. Issues linked below:
+- Afterburn [[1](https://issues.redhat.com/browse/OCPBUGS-7559)],[[2](https://issues.redhat.com/browse/OCPBUGS-4769)]
+- podman [[1](https://issues.redhat.com/browse/OCPBUGS-9969)]
+- skopeo [[1](https://issues.redhat.com/browse/OCPBUGS-3621)]
+
+Additionally, the stub secret [referenced](https://github.com/openshift/installer/blob/1ca0848f0f8b2ca9758493afa26bf43ebcd70410/pkg/asset/machines/gcp/machines.go#L197) in the `MachineSet` is also not managed. This stub is used by the ignition binary in firstboot to auth and consume content from the `machine-config-server`(MCS). The content served includes the actual ignition configuration and the target OCI format RHCOS image. The ignition binary now does first boot provisioning based on this, then hands off to the `machine-config-daemon`(MCD) first boot service to do the reboot into the target OCI format RHCOS image. As 4.6 and up clusters only understood spec 3 ignition, and as the unmanaged ignition stub is only spec 2, this was now an incompatibility. This would prevent new nodes from joining a cluster that had been upgraded past 4.5, but was originally a 4.5 or lower at install time. 
```

Comment:
You're looking at managing user data secrets. These may be shared between many machinesets. You are allowing the user to select a subset of MachineSets to manage. What strategy will you employ to ensure that all MachineSets using a user data secret continue to work if some are opted-in, and some are opted-out of the update functionality introduced here?

Reply:
My line of thinking was, during the sync loop for an enrolled machineset:
- check the stub secret referenced in the providerSpec, if it spec 3, we don't need to do anything
- if spec 2, create a new secret which up-translates the old stub secret
- update the providerSpec to reference the new ignition stub

So if a machineset is not enrolled, it should still be able to reference the older stub, as it still exists.

Reply:
One more thing to note is that, I'd have to check if an upgraded spec 3 version of the secret exists, before I attempt to update a spec 2, just in case we've already updated the same secret for a different machineset.

Reply:
To be able to check this, you need predictable names, can you define a naming convention for the v2 to v3 converted secrets?

Reply:
The MCO today kinda has 2 already: worker-user-data-managed and master-user-data-managed.

We could maybe leverage those for existing pools and have a similar naming mechanism for the new ones?

But that's a good point on the some-updated-some-not split. I was originally thinking we could just update secrets inplace but I think we should generate some new managed-ones

Reply:
If the existing managed cases wouldn't break by updating them, then that could be used, else yes, a new managed spec 3 name or something sounds good

Reply:
Me and Jerry chatted about the approach here a bit and I had some follow-up questions/concerns. I thought it best to mention them here as it is easier to track than elsewhere.

(i) The existing `*-managed` secrets are synced by the MCO and are [generated by the operator](https://github.com/openshift/machine-config-operator/blob/795ad9d59e2cc505f1a4c878f2a8928f3ec5d7b6/pkg/controller/common/helpers.go#L211). They do not react to changes to the original stub secret and hence will not include any user customization. For most clusters, we should be safe to use these. Would there be any new ones generated if the user were to create any custom pools? Is the only con with this approach losing user customization?

(ii) Our original plan was to create a new spec 3 secret(`*-spec-3-managed`) from the original stub secret and let it co-exist with the already implemented `*-managed`. While this approach will keep user customization, these uptranslated secrets could potentially contain out of date certs, as `MSBIC` is not really actively managing/syncing them. Is this the main concern with this approach here, or is there something else?

Considering both of these scenarios, I am wondering if we should go with (i), while providing a disclaimer about them potentially losing their customization if they were turn this feature on.  Or perhaps the let the user override their stub secret, but warn them that their TLS cert could potentially be out of date in the future by taking the wheel?

Reply:
> Would there be any new ones generated if the user were to create any custom pools?

No (and to a degree, you shouldn't have created machinesets for custom pools, since we don't support booting into them directly. Some customers have done this though)

> Is the only con with this approach losing user customization?

Eventually we can make a copy of what was in the other stub secret, but yes, if we just use this fresh, we will lose user customizations by default

> Is this the main concern with this approach here, or is there something else?

I think a most clusters are probably post-4.6 installed, so this might not be a big concern. If certs are out of date, it wouldn't have worked for the original case anyways, so I don't see too many issues

> I am wondering if we should go with (i), while providing a disclaimer about them potentially losing their customization if they were turn this feature on

I think that's a fine place to start. Eventually we can migrate the user to some new API to customize their stub secrets. For 99.9% of cases, users don't touch the stub secret, so as long as we leave an eventual out (or a documented workaround) we should be fine

Reply:
Makes sense. I'll update the enhancement to reflect this approach then:
- Point to the existing managed stubs for enrolled machinesets, but have a disclaimer in the documentation about losing user customization when turning on boot image updates
- Outline that a way to support user customization will be added, either via API or doc, in a future release